# 1. Logistic Regression

(a) 
> (10 points) Do natural log transform of the PFOS variable in file
pfas.csv and store the results as a new variable log PFOS in the
data file. Standardize the variables x=[log PFOS, age, gender, BMI].

In [2]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler
import statsmodels.formula.api as api
import statsmodels

C:\Users\brean\AppData\Local\Temp\ipykernel_28160\3504629728.py:1: DeprecationWarning: 
Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),
(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)
but was not found to be installed on your system.
If this would cause problems for you,
please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466
        
  import pandas as pd


In [3]:
def autoscale(df, variables_to_scale):
    """standardizes variables"""
    for variable in variables_to_scale:
        df[variable] = ( df[variable] - np.mean(df[variable]) ) / np.std(df[variable])



In [4]:
df = pd.read_csv("pfas.csv")
df["log_PFOS"] = np.log(df["PFOS"])

autoscale(df, ["age", "gender", "BMI", "log_PFOS"])

#### Initalize variables
Y = df["disease"].to_numpy()
X = df.loc[:,["age", "gender", "BMI", "log_PFOS"]].to_numpy()
## add intercept
X_with_intercept = np.insert( X, 0, np.array([1]*X.shape[0]), axis = 1 )


print("disease\n", Y)
print("     intercept|    age   |   gender     |  BMI    |    log_PFOS\n",X_with_intercept)

disease
 [0 0 0 0 0 0 1 1 1 1 1 1 1 0 1 0 1 1 1 1 0 1 0 1 1 0 0 1 0 1 0 0 0 0 1 0 1
 1 1 0 0 0 0 0 0 1 0 1 1 1 1 1 0 1 0 1 0 1 1 1 0 1 1 1 1 0 0 1 1 0 1 1 0 0
 0 1 1 1 1 1 1 0 1 0 0 1 1 0 1 0 1 0 0 1 0 1 0 0 0 0 0 1 1 1 1 0 0 1 1 1 0
 0 1 0 1 1 1 1 1 0 0 0 1 0 0 1 1 1 0 0 0 1 1 0 1 1 1 1 1 0 0 1 1 0 1 0 1 1
 0 1 1 0 1 0 0 0 1 0 1 1 0 1 0 1 0 0 1 1 0 0 1 1 0 1 0 1 0 0 1 0 1 1 0 1 1
 0 1 1 1 1 0 0 0 0 0 1 1 0 1 0 1 0 0 1 1 0 1 1 1 1 0 1 0 0 0 1 1 0 1 0 1 1
 1 1 1 0 0 0 1 1 1 0 1 0 0 0 0 0 0 1 0 0 1 0 0 1 1 0 1 0 1 1 0 0 0 1 1 0 0
 1 1 0 0 1 0 0 0 0 1 0 1 0 1 1 1 0 1 1 0 0 1 1 1 1 1 0 1 1 1 1 1 1 0 1 0 1
 0 0 1 0]
     intercept|    age   |   gender     |  BMI    |    log_PFOS
 [[ 1.         -1.06321047 -1.06191317 -0.78453458  0.90375231]
 [ 1.          0.96375408 -1.06191317 -1.18985472 -0.99778188]
 [ 1.          0.31512542  0.94169658 -0.5470809  -0.24350772]
 ...
 [ 1.          0.396204    0.94169658  0.07829738 -0.24248557]
 [ 1.          0.15296826  0.94169658  0.8139823  -0.231952

#### Initalize variables using StandardScaler

In [297]:
df2 = pd.read_csv("pfas.csv")
df2["log_PFOS"] = np.log2(df2["PFOS"])

# StandardScaler().fit(df2[["log_PFOS", "age", "gender", "BMI"]])
X = StandardScaler().fit_transform(df2[["age", "gender", "BMI", "log_PFOS"]])
X_with_intercept2 = np.insert( X, 0, np.array([1]*X.shape[0]), axis = 1 )

Y = df["disease"].to_numpy()

print("disease\n", Y)
print("     intercept|    age   |   gender     |  BMI    |    log_PFOS\n",X_with_intercept2)


disease
 [0 0 0 0 0 0 1 1 1 1 1 1 1 0 1 0 1 1 1 1 0 1 0 1 1 0 0 1 0 1 0 0 0 0 1 0 1
 1 1 0 0 0 0 0 0 1 0 1 1 1 1 1 0 1 0 1 0 1 1 1 0 1 1 1 1 0 0 1 1 0 1 1 0 0
 0 1 1 1 1 1 1 0 1 0 0 1 1 0 1 0 1 0 0 1 0 1 0 0 0 0 0 1 1 1 1 0 0 1 1 1 0
 0 1 0 1 1 1 1 1 0 0 0 1 0 0 1 1 1 0 0 0 1 1 0 1 1 1 1 1 0 0 1 1 0 1 0 1 1
 0 1 1 0 1 0 0 0 1 0 1 1 0 1 0 1 0 0 1 1 0 0 1 1 0 1 0 1 0 0 1 0 1 1 0 1 1
 0 1 1 1 1 0 0 0 0 0 1 1 0 1 0 1 0 0 1 1 0 1 1 1 1 0 1 0 0 0 1 1 0 1 0 1 1
 1 1 1 0 0 0 1 1 1 0 1 0 0 0 0 0 0 1 0 0 1 0 0 1 1 0 1 0 1 1 0 0 0 1 1 0 0
 1 1 0 0 1 0 0 0 0 1 0 1 0 1 1 1 0 1 1 0 0 1 1 1 1 1 0 1 1 1 1 1 1 0 1 0 1
 0 0 1 0]
     intercept|    age   |   gender     |  BMI    |    log_PFOS
 [[ 1.         -1.06321047 -1.06191317 -0.78453458  0.90375231]
 [ 1.          0.96375408 -1.06191317 -1.18985472 -0.99778188]
 [ 1.          0.31512542  0.94169658 -0.5470809  -0.24350772]
 ...
 [ 1.          0.396204    0.94169658  0.07829738 -0.24248557]
 [ 1.          0.15296826  0.94169658  0.8139823  -0.231952

In [298]:
False in (X_with_intercept == X_with_intercept2)

False

### Numerical Solution
(b) 
> (35 points) Use y=disease and the standardized x=[PFOS, age,
gender, BMI] to write and debug your own gradient descent algorithm for logistic regression. Your algorithm should export the
learned parameters in the θ vector. Note that you can modify the
gradient descent algorithm that you have written for the linear regression algorithm to achieve logistic regression.

#### Sigmoid

$ g(B^Tx)= \frac{1}{1+e^{-(B^Tx)}}$

$B^Tx=B_0x_0+B_1x_1+...+B_nx_n$

$x_0=1$


$$
y = \begin{cases}
    \text{1 if} & B^Tx\text{ ≥ 0} \\
    \text{0 if} & B^Tx\text{< 0} \\
\end{cases}
$$


#### Cost Function

$i$ is a row $j$ is a column

$ J(B) = -\frac{1}{m} \sum_{i=1}^m [y_i ln(\frac{1}{1+e^{-(B^Tx)}}) + (1-y_i)ln(\frac{1}{1+e^{-(B^Tx)}})]$

$ B_j := B_j-α(\frac{∂J(B)}{∂B_j}) $

$ \frac{∂J(B)}{∂B_j} = \frac{1}{m} \sum [(\frac{1}{1+e^{-(B^Tx)}}) - y_i]x_{j,i}$



In [5]:
LEARNING_RATE = 0.1
EPOCHS = 1000

b_vector = np.zeros(X_with_intercept.shape[1]) ## initilize intercept and coefficents as 0
cols = ["intercept","log_PFOS", "age", "gender", "BMI"]


def compute_sigmoid(B: np.ndarray, X: np.ndarray):

    # return 1 / (1 + np.exp(-np.dot(X, B))) ## ALT METHOD: np.dot(X, B) computes the dot product between each row of X and B
    return 1/(  1 + np.exp( - X @ B )  ) ## X @ B computes the dot product between each row of X and B


def compute_cost(B: np.ndarray, X: np.ndarray, Y: np.ndarray):
    num_rows = Y.shape[0]
    sigmoid = compute_sigmoid(B, X)

    return np.sum((Y * np.log(sigmoid)) + (1 - Y) * np.log(sigmoid)) / -num_rows

def partial_derivative(B: np.ndarray, X: np.ndarray, Y: np.ndarray):
    num_rows = Y.shape[0]
    sigmoid = compute_sigmoid(B, X)

    # return (1 / num_rows) * np.dot(X.T, (sigmoid - Y)) ## ALT METHOD
    
    ## (sigmoid - Y)[:,np.newaxis] changes its shape from (n,) to (n, 1)
    return np.sum( (sigmoid - Y)[:,np.newaxis] * X, axis = 0 ) / num_rows

for epoch in range(EPOCHS):

    cost = compute_cost(b_vector, X_with_intercept, Y)
    
    b_vector -= LEARNING_RATE * partial_derivative(b_vector, X_with_intercept, Y)

    b_dict = {col : b for col, b in zip(cols, b_vector)}

    if epoch % 100 == 0 or epoch+1 == 1000:
        print(f"------------------------------------epoch: {epoch}------------------------------------")
        print(f"\t\t\t\tcost: {cost}\nb_dict: {b_dict}")


print("final values: ")
for key, value in b_dict.items(): print(f"{key}: {round(value,8)}")


------------------------------------epoch: 0------------------------------------
				cost: 0.6931471805599453
b_dict: {'intercept': 0.003666666666666667, 'log_PFOS': 0.005614781879612744, 'age': 0.0071261720081734915, 'gender': 0.009237937363079844, 'BMI': 9.976846641748035e-05}
------------------------------------epoch: 100------------------------------------
				cost: 0.6539572177431865
b_dict: {'intercept': 0.13972505594387125, 'log_PFOS': 0.21388705051948634, 'age': 0.2507497017915419, 'gender': 0.3237468146185928, 'BMI': 0.023679711181707166}
------------------------------------epoch: 200------------------------------------
				cost: 0.6522490102240123
b_dict: {'intercept': 0.15402810805273662, 'log_PFOS': 0.23758336393223678, 'age': 0.2717541566678402, 'gender': 0.34957447968201444, 'BMI': 0.031900536683440256}
------------------------------------epoch: 300------------------------------------
				cost: 0.6520341102384797
b_dict: {'intercept': 0.15567719833985094, 'log_PFOS': 0.24

(c) 
> (10 points) Apply your own algorithm to the standardized data and
provide the values of the learned θ.

In [ ]:
## for each row (sample) in x, multiply the variables (x1-x4) by the weights in b
## sum the resulting vector 
## this is the dot product 

print("final intercept and coefficents: ")
for key, value in b_dict.items(): print(f"{key}: {round(value,8)}")

y_predicted = np.dot( X_with_intercept, b_vector )
# y_predicted = np.dot( X, b_vector[1: len(b_vector)] ) + b_vector[0] ## b_vector[0] is the intercept

print("\npredicted y values:\n", y_predicted )


final intercept and coefficents: 
intercept: 0.15589359
log_PFOS: 0.24116572
age: 0.27433772
gender: 0.35218456
BMI: 0.03370199

predicted y values:
 [-0.63768188 -0.35568051  0.28935379 -0.71980975 -0.22064997  1.35101711
 -0.93051684 -0.18920002 -0.26944778  0.31042177  0.28683218 -0.02217559
  0.98501291  0.72088543 -0.21933665 -0.85168726  0.99216161  1.02343907
  1.0547744  -0.26340074  0.41901037  0.22533793  0.01916788  0.2273552
 -0.87785937  0.3756261   0.6163954  -0.42789026 -0.09631984 -0.83017099
 -0.67437333 -0.50946058 -0.20149582 -0.24332532  0.12917514  0.40945855
 -0.14580022  0.90564431  0.21993656  0.39186022 -0.513669   -0.83247607
  0.02116505 -0.12098017  0.74917571 -0.02371777  1.22138296  0.12075989
  0.94379943  0.91592837 -0.0973883   0.45756781  0.79816807 -0.54525029
  0.11661924  0.36770702 -0.7848109  -0.21367765 -0.16292993 -0.06461848
 -1.33210208  1.07282734  0.66071003  0.22931989  1.09987897  0.95323906
  0.63038586  0.1478374  -0.47250965 -0.16040242

In [205]:
diff = np.where(y_predicted > 0, 1, 0) - Y
len(diff[diff > 0]) / 300 ## 78 missclassified

0.26

### Sklearn

(d) 
> (10 points) Apply LogisticRegression in sklearn to the y and the stan-
dardized x. What are the θ values you get from sklearn? Information
about how to apply LogisticRegression in sklear can be found at
https://scikit-learn.org/stable/modules/generated/sklearn.
linear_model.LogisticRegression.html

In [11]:

logreg = LogisticRegression(random_state=16,  max_iter=1000, penalty=None)

# model = logreg.fit(X_df, Y_df)
SklearnModel = logreg.fit(X, Y)

coef_df = {col : coef for col, coef in zip(list(cols[1:len(cols)]), SklearnModel.coef_[0])}
# print(coef_df)
# print(X.columns)
print(f"coefficent: {coef_df}\nintercept: {SklearnModel.intercept_}")



coefficent: {'log_PFOS': 0.2415679442838867, 'age': 0.27431454819456, 'gender': 0.3523996420149074, 'BMI': 0.033675373184102346}
intercept: [0.1560249]


### statsmodel
(e) 
> (10 points) Add constant to the standardized x using the function
add constant. Instructions about how to use add constant can be
found at:
https://www.statsmodels.org/dev/generated/statsmodels.tools.tools.add_constant.html
Apply Logit in statsmodels to the data with constant 1 added. What
θ do you get? Instructions about how to use statsmodels to do logistic
regression can be found at:
https://www.statsmodels.org/stable/generated/statsmodels.formula.api.logit.html

In [26]:
X_df = df.loc[:,["age", "gender", "BMI", "log_PFOS"]]

Y_df = df["disease"]

X_df_with_intercept = statsmodels.tools.tools.add_constant(X_df, prepend=True)
model = api.logit(formula="disease ~ age + gender + BMI + log_PFOS", data = pd.concat([X_df_with_intercept, Y_df], axis = 1))

model.fit().summary()


Optimization terminated successfully.
         Current function value: 0.658576
         Iterations 5


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                disease   No. Observations:                  300
Model:                          Logit   Df Residuals:                      295
Method:                           MLE   Df Model:                            4
Date:                Thu, 02 Oct 2025   Pseudo R-squ.:                 0.04617
Time:                        18:09:06   Log-Likelihood:                -197.57
converged:                       True   LL-Null:                       -207.14
Covariance Type:            nonrobust   LLR p-value:                 0.0007418
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.1559      0.120      1.303      0.193      -0.079       0.390
age            0.2412      0.123      1.963      0.050       0.000       0.482
gender         0.2743      0.120      2.280      0.023       0.038       0.510
BMI            0.3522      0.123      2.868      0.004       0.112       0.593
log_PFOS       0.0337      0.121      0.280      0.780      -0.203       0.270
==============================================================================
"""

> (f) (25 points) Compare θ from your own algorithm, θ from LogisticRe-
gression in sklearn, and θ from statsmodel. Do you get very similar
results? If not, what could you do to make the θ values similar?

Below are the coefficents and intercepts for all of the models. 

**My final intercept and coefficents:**
|Coefficent   | Value  |
|-------------|--------|
|intercept    |0.155894|
|log_PFOS     |0.241166|
|age          |0.274338|
|gender       |0.352185|
|BMI          |0.033702|
-----------------------------------------------------------------
**Sklearn intercept and coefficents:**
|Coefficent   | Value  |
|-------------|--------|
|intercept    |0.156025|
|log_PFOS     |0.241568|
|age          |0.274315|
|gender       |0.3524  |
|BMI          |0.033675|
-----------------------------------------------------------------
**StatsModel intercept and coefficents:**
|Coefficent   | Value  |
|-------------|--------|
|Intercept    |0.155894|
|age          |0.241166|
|gender       |0.274338|
|BMI          |0.352185|
|log_PFOS     |0.033702|

The coefficents from sklearn and statsmodel were practically the same as my values which means 

In [ ]:
print("My final intercept and coefficents:")
for key, value in b_dict.items(): print(f"{key}: {round(value, 6)}")

print("-----------------------------------------------------------------")
print("Sklearn intercept and coefficents:")
print(f"intercept: {round(SklearnModel.intercept_[0], 6)}")
for key, value in coef_df.items(): print(f"{key}: {round(value, 6)}")
print("-----------------------------------------------------------------")
print("StatsModel intercept and coefficents:\n")
model.fit().params


My final intercept and coefficents: 
intercept: 0.155894
log_PFOS: 0.241166
age: 0.274338
gender: 0.352185
BMI: 0.033702
-----------------------------------------------------------------
Sklearn intercept and coefficents
intercept: 0.156025
log_PFOS: 0.241568
age: 0.274315
gender: 0.3524
BMI: 0.033675
-----------------------------------------------------------------
StatsModel intercept and coefficents

Optimization terminated successfully.
         Current function value: 0.658576
         Iterations 5


Intercept    0.155894
age          0.241166
gender       0.274338
BMI          0.352185
log_PFOS     0.033702
dtype: float64